packages/@aws-cdk/aws-ecs/lib/cluster.ts

```diff
@@ -105,19 +115,29 @@ export class Cluster extends Resource implements ICluster {
     });
 
     const containerInsights = props.containerInsights !== undefined ? props.containerInsights : false;
-    const clusterSettings = containerInsights ? [{ name: 'containerInsights', value: 'enabled' }] : undefined;
+    const clusterSettings = containerInsights ? [{name: 'containerInsights', value: 'enabled'}] : undefined;
+    this.clusterName = props.clusterName ?? `${Stack.of(this).stackName}-${id}`
```

Comment:
I am assigning the `this.clusterName` here rather than auto-assigned by cloudformation so we can simply use `this.clusterName` in UserData without dependency on the cluster construct. Previously the `clusterName` came from `this.physicalName` and we will not resolve this until the resource is created, in which case the UserData will depend on the cluster being created.

This will cause breaking changes and might force the cluster re-creation. But I believe this should be a better approach.